feat: integrate & test FIP-0098 additions

rvagg opened the pull request:
Ref: https://github.com/filecoin-project/go-state-types/pull/376
Ref: https://github.com/filecoin-project/builtin-actors/pull/1639

I've added some extras in our state types abstractions here, I'd like to add more and reduce (ideally remove) the direct uses of go-state-types versioned actor types we have through the codebase. (See https://github.com/filecoin-project/lotus/pull/12960 for some of how annoying this is). My goal in the itest was to not refer to go-state-types at all while testing all of those additions. We should be able to approach almost all other uses of those types throughout the code in the same way, but we're pretty sloppy.

Also removed some batch balancer stuff.

BigLep commented:
@rvagg : I assume this should be backported?

@rjan90 : would you ideally want to get this merged before the calibnet upgrade so migration pathways are flexed with this code?  (I assume we'd just encourage a few calibnodes to use `master` with this change rather than cutting another RC at this point.)

Stebalien commented on chain/actors/builtin/miner/actor.go.template:
```diff
@@ -319,3 +332,37 @@ func AllCodes() []cid.Cid {
+	case actorstypes.Version{{.}}:
+		{{- if (ge . 16)}}
+			return minertypes{{.}}.PledgePenaltyForContinuedFault(
+				smoothing{{.}}.FilterEstimate{
```
later: we should consider moving `FilterEstimate` to a shared package.

rvagg replied:
yep, on my list

Stebalien commented on chain/actors/builtin/miner/actor.go.template:
```diff
@@ -319,3 +332,37 @@ func AllCodes() []cid.Cid {
+				qaSectorPower,
+			), nil
+		{{- else}}
+				return big.Zero(), nil
```
This is a lie.

Stebalien replied:
Honestly, the pledge penalty is the same for all older versions. We could just call into v16 for all versions.

rvagg replied:
great, cause I didn't want to backport the function to all the previous versions

Stebalien reviewed (approved):
One suggestion but otherwise this looks correct.

rvagg commented:
I also added a `miner.PledgePenaltyForTermination` utility wrapper for the v16 implementation while I was in there, but this one will error for everything prior to v16 since that's the simplified one we introduced.

Stebalien commented on chain/actors/builtin/miner/actor.go.template:
```diff
@@ -334,35 +336,75 @@ func AllCodes() []cid.Cid {
+        return big.Zero(), err
+    }
+
+    if v < actorstypes.Version16 {
+        return big.Zero(), xerrors.Errorf("not supported for network version %d", v)
+    }
```
Is this necessary? We already have a default below.

rvagg replied:
I guess not, I was imagining the different language would be more helpful than it actually is

removed

rvagg commented:
I shouldn't have merged this without a new go-state-types, I'll get that updated now in a new PR
